lightning/src/ln/channelmanager.rs

```diff
@@ -3472,16 +3498,45 @@ where
 	/// [`events::Event::PaymentClaimed`] events even for payments you intend to fail, especially on
 	/// startup during which time claims that were in-progress at shutdown may be replayed.
 	pub fn fail_htlc_backwards(&self, payment_hash: &PaymentHash) {
+		self.fail_htlc_backwards_with_reason(payment_hash, &FailureCode::IncorrectOrUnknownPaymentDetails);
+	}
+
+	/// This is a variant of [`ChannelManager::fail_htlc_backwards`] that allows you to specify the
+	/// reason for the failure.
+	///
+	/// See [`FailureCode`] for valid failure codes.
+	pub fn fail_htlc_backwards_with_reason(&self, payment_hash: &PaymentHash, failure_code: &FailureCode) {
 		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(&self.total_consistency_lock, &self.persistence_notifier);
 
 		let removed_source = self.claimable_payments.lock().unwrap().claimable_htlcs.remove(payment_hash);
 		if let Some((_, mut sources)) = removed_source {
 			for htlc in sources.drain(..) {
-				let mut htlc_msat_height_data = htlc.value.to_be_bytes().to_vec();
-				htlc_msat_height_data.extend_from_slice(&self.best_block.read().unwrap().height().to_be_bytes());
-				let source = HTLCSource::PreviousHopData(htlc.prev_hop);
-				let reason = HTLCFailReason::reason(0x4000 | 15, htlc_msat_height_data);
+				let source = HTLCSource::PreviousHopData(htlc.prev_hop.clone());
 				let receiver = HTLCDestination::FailedPayment { payment_hash: *payment_hash };
+
+				let reason = match failure_code {
+					FailureCode::TemporaryNodeFailure => HTLCFailReason::from_failure_code(*failure_code as u16),
+					FailureCode::RequiredNodeFeatureMissing => HTLCFailReason::from_failure_code(*failure_code as u16),
+					FailureCode::ExpiryTooSoon => {
+						let short_channel_id = htlc.prev_hop.short_channel_id;
+						let (counterparty_node_id, chan_id) = self.short_to_chan_info.read().unwrap().get(&short_channel_id).cloned().unwrap();
+						let per_peer_state = self.per_peer_state.read().unwrap();
+						let peer_state = per_peer_state.get(&counterparty_node_id).unwrap().lock().unwrap();
+						let chan = peer_state.channel_by_id.get(&chan_id).unwrap();
```

Comment:
Sadly neither of these `unwrap`s are safe.. I think we should default to the normal `IncorrectOrUnknownPaymentDetails` if either are `None`

Reply:
Just changed to handle the `unwrap`s from the `get`s. Just wanted to make sure that's what you meant right, and not also the `unwrap` after the `lock`?

Reply:
Oops my bad! Will fix those right now

Reply:
I think I got the other ones--if I've checked everything right the only unwraps should be from locks now.

Also would it be valuable for me test these specific paths where it's defaulting? And if so are there any existing tests (or general areas) you could point me to to see how I might simulate these situations?